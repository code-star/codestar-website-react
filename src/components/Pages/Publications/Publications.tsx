import React, { FC } from 'react';
import { translate, TranslationFunction } from 'react-i18next';
import {Fade, Slide, Typography} from '@material-ui/core';
import compose from 'recompose/compose';
import Container from '../../../Container/Container';
import Section from '../../Molecules/Section/Section';
import { withStyles } from '@material-ui/core/styles';
import Publications from '../../Molecules/Publications/Publications';
import { styles } from '../About/About.styles';

interface IPropsInner {
  classes: any;
  t: TranslationFunction;
}

interface IPropsOuter {}

export const PublicationsPage: FC<IPropsInner> = ({ t, classes }) => {
  const [publicationsText0, publicationsText1] = t('PUBLICATIONS_TEXTS', {
    returnObjects: true,
  });
  return (
    <Section scrollname={'publications'}>
      <Container marginTopNavBar>
        <div className="row justify-content-center">
          <div className={`col-12 col-md-8 ${classes.text}`}>
            <Fade in timeout={2000}>
              <div className="mb-4">
                <Typography variant="h2" color="inherit" gutterBottom>
                  {t('PUBLICATIONS_TITLE')}
                </Typography>
                <Typography variant="body1" color="inherit" gutterBottom>
                  {publicationsText0}
                  <a href="https://youtube.com/codestar" className={classes.link}>Codestar YouTube</a>
                  {publicationsText1}
                </Typography>
              </div>
            </Fade>
            <Slide in timeout={1000} direction="up">
              <div>
                <Publications />
              </div>
            </Slide>
          </div>
        </div>
      </Container>
    </Section>
  );
};

export default compose<IPropsInner, IPropsOuter>(
  withStyles(styles),
  translate(['about'], { wait: true })
)(PublicationsPage);
