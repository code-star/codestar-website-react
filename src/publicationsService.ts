export interface IPublication {
  id: string,
  latestPublishedAt: string,
  author: string,
  authorImg: string,
  title: string,
  paragraphs: string[],
  uniqueSlug: string,
  previewImgId: string
}

let cachedPublications: IPublication[] = [];

// This has been wrapped in a function to able to run unit tests where process.env.REACT_APP_STAGE is changed
function getUrl(lambdaName: string): string {
  if (process.env.REACT_APP_STAGE === 'dev') {
    return `/mock/${lambdaName}.json`;
  }
  const AWS_PREFIX =
    process.env.REACT_APP_STAGE === 'test' ? 'hjoutysc5k' : '267sder6c7';
  const AWS_STAGE = process.env.REACT_APP_STAGE === 'test' ? 'test' : 'prod';
  return `https://${AWS_PREFIX}.execute-api.eu-west-1.amazonaws.com/${AWS_STAGE}/${lambdaName}`;
}

async function fetchPublications(): Promise<IPublication[]> {
  try {
    const url = getUrl('get-publications');
    cachedPublications = await fetch(url).then(data => data.json());
    return cachedPublications;
  } catch (err) {
    // fail silently
    return [];
  }
}

// Should be drop-in replaceable with Redux Thunk
export async function getCachedPublications(): Promise<IPublication[]> {
  return cachedPublications.length
    ? cachedPublications
    : fetchPublications();
}
