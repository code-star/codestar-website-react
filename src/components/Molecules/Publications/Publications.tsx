import React, { FC } from 'react';
import { translate, TranslationFunction } from 'react-i18next';
import compose from 'recompose/compose';
import { lifecycle } from 'recompose';
import PublicationCard from '../PublicationCard/PublicationCard';
import { getCachedPublications, IPublication } from "../../../publicationsService";

type PropsInner = {
  t: TranslationFunction;
  publications: IPublication[];
}

type PropsOuter = {}

type Props = PropsInner & PropsOuter;

// TODO intro text above publications incl translations
// TODO combine Medium from API and other publications from static JSON
// TODO /blog

export const Publications: FC<Props> = ({publications = []}) => {
  const publicationCards = publications.map((p) => <PublicationCard key={p.id} publication={p}></PublicationCard>);
  return (<>{publicationCards}</>);
};

const withUserData = lifecycle({
  async componentDidMount() {
    try {
      const publications = await getCachedPublications();
      this.setState({ publications });
    } catch (err) {
      // TODO
      console.error(err);
    }
  }
});

export default compose<Props, PropsOuter>(
  withUserData,
  translate(['about'], { wait: true })
)(Publications);
