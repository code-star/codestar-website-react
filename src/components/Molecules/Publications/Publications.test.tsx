import { Publications } from './Publications';
import { IPublication } from '../../../publicationsService';
import renderer from "react-test-renderer";
import React from "react";

jest.mock('../PublicationCard/PublicationCard', () => () => "mock-publication-card");

const publications: IPublication[] = [{
  id: '1',
  latestPublishedAt: '100000000',
  author: 'Test Author',
  authorImg: 'foo.png',
  title: 'Test Title',
  paragraphs: ['Test Paragraph'],
  uniqueSlug: 'slug',
  previewImgId: 'preview.png',
}];

describe('<Publications />', () => {
  it('renders publications', () => {
    const comp = renderer
      .create(
        <Publications key={10} publications={publications} />
      )
      .toJSON();
    expect(comp).toMatchSnapshot();
  });
});
