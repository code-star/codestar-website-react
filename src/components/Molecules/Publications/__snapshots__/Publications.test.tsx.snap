// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Publications /> renders publications 1`] = `"mock-publication-card"`;
